### DISCUSSION TITLE
unable to build web-vault

### DISCUSSION THREAD

**Discussion by rapkin61** [dani-garcia/vaultwarden#1671] [Q&A]
**git clone --recurse-submodules https://github.com/bitwarden/web.git web-vault**
Cloning into 'web-vault'...
remote: Enumerating objects: 26356, done.
remote: Counting objects: 100% (940/940), done.
remote: Compressing objects: 100% (437/437), done.
remote: Total 26356 (delta 597), reused 775 (delta 485), pack-reused 25416
Receiving objects: 100% (26356/26356), 132.85 MiB | 5.99 MiB/s, done.
Resolving deltas: 100% (19172/19172), done.
Submodule 'jslib' (https://github.com/bitwarden/jslib.git) registered for path 'jslib'
Cloning into '/home/klaus/web-vault/jslib'...
remote: Enumerating objects: 10753, done.        
remote: Counting objects: 100% (1253/1253), done.        
remote: Compressing objects: 100% (641/641), done.        
remote: Total 10753 (delta 724), reused 1029 (delta 574), pack-reused 9500        
Receiving objects: 100% (10753/10753), 3.42 MiB | 5.69 MiB/s, done.
Resolving deltas: 100% (7805/7805), done.
Submodule path 'jslib': checked out '4eb50d757d03343ee842ea68a6b2c8282ef3f382'
**cd web-vault/**
**git checkout "$(git tag --sort=v:refname | tail -n1)"**
M	jslib
Note: checking out 'v2.19.0'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 9d42a565 update jslib

....so ... this is v2.19 - we need to apply patch for that

**git apply v2.19.0.patch** 

**npm install**

 node-sass@4.13.1 install /home/klaus/web-vault/node_modules/node-sass
 node scripts/install.js

Downloading binary from https://github.com/sass/node-sass/releases/download/v4.13.1/linux-x64-67_binding.node
Download complete .] - :
Binary saved to /home/klaus/web-vault/node_modules/node-sass/vendor/linux-x64-67/binding.node
Caching binary to /home/klaus/.npm/node-sass/4.13.1/linux-x64-67_binding.node

 ngx-infinite-scroll@7.0.1 postinstall /home/klaus/web-vault/node_modules/ngx-infinite-scroll
 opencollective postinstall

                                          Thanks for installing ngx-infinite-scroll 
                                        Please consider donating to our open collective
                                               to help us maintain this package.

                                                   Number of contributors: 0
                                                     Number of backers: 7
                                                     Annual budget: US$ 29
                                                   Current balance: US$ 189

                             Donate: https://opencollective.com/ngx-infinite-scroll/donate

 node-sass@4.13.1 postinstall /home/klaus/web-vault/node_modules/node-sass
 node scripts/build.js

Binary found at /home/klaus/web-vault/node_modules/node-sass/vendor/linux-x64-67/binding.node
Testing binary
Binary is fine

 webpack-cli@3.2.1 postinstall /home/klaus/web-vault/node_modules/webpack-cli
 lightercollective

     *** Thank you for using webpack-cli! ***

Please consider donating to our open collective
     to help us maintain this package.

  https://opencollective.com/webpack/donate

                    ***

 bitwarden-web@2.19.0 postinstall /home/klaus/web-vault
 npm run sub:init

 bitwarden-web@2.19.0 sub:init /home/klaus/web-vault
 git submodule update --init --recursive

error: Your local changes to the following files would be overwritten by checkout:
	src/angular/components/register.component.ts
Please commit your changes or stash them before you switch branches.
Aborting
Unable to checkout 'f80e89465ffc004705d2941301c0ffb6bfd71d1a' in submodule path 'jslib'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! bitwarden-web@2.19.0 sub:init: `git submodule update --init --recursive`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the bitwarden-web@2.19.0 sub:init script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/klaus/.npm/_logs/2021-05-03T09_24_53_641Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! bitwarden-web@2.19.0 postinstall: `npm run sub:init`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the bitwarden-web@2.19.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/klaus/.npm/_logs/2021-05-03T09_24_53_671Z-debug.log

**Comment by rapkin61** [comment]
...excuse me for that strange kind of formatting!
How can I switch that off?

**Reply by mqus** [reply]
you could try to format it as source code by putting a ` ``` ` in the line before and after the log. But it could be possible that there are parts that don't work with that, try the preview and format yourself if you need :)

**Comment by Arathen** [comment]
This is a +1 for me as well. Im also getting the exact same error when trying to build the web vault.

```
arathen@ubuntu98 ~/src/web-vault $ npm install

  bitwarden-web@2.19.0 postinstall /home/arathen/src/web-vault
  npm run sub:init


  bitwarden-web@2.19.0 sub:init /home/arathen/src/web-vault
  git submodule update --init --recursive

error: Your local changes to the following files would be overwritten by checkout:
        src/angular/components/register.component.ts
Please commit your changes or stash them before you switch branches.
Aborting
Unable to checkout 'f80e89465ffc004705d2941301c0ffb6bfd71d1a' in submodule path 'jslib'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! bitwarden-web@2.19.0 sub:init: git submodule update --init --recursive
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the bitwarden-web@2.19.0 sub:init script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/arathen/.npm/_logs/2021-05-04T22_25_28_504Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! bitwarden-web@2.19.0 postinstall: npm run sub:init
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the bitwarden-web@2.19.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/arathen/.npm/_logs/2021-05-04T22_25_28_979Z-debug.log

arathen@ubuntu98 ~/src/web-vault $ git status
HEAD detached at v2.19.0
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   jslib (new commits, modified content)
        modified:   package-lock.json
        modified:   src/404.html
        modified:   src/app/app.component.ts
        modified:   src/app/layouts/footer.component.html
        modified:   src/app/layouts/frontend-layout.component.html
        modified:   src/app/layouts/navbar.component.html
        modified:   src/app/organizations/settings/organization-subscription.component.ts
        modified:   src/app/send/access.component.html
        modified:   src/app/services/services.module.ts
        modified:   src/app/settings/two-factor-u2f.component.ts
        modified:   src/scss/styles.scss

no changes added to commit (use "git add" and/or "git commit -a")

arathen@ubuntu98 ~/src/web-vault $ git show
commit 9d42a565dbef29c5c361dbe5b3db88d765d65a98 (HEAD, tag: v2.19.0)
Author: Kyle Spearrin <kyle.spearrin@gmail.com>
Date:   Thu Mar 11 10:49:49 2021 -0500

  update jslib

diff --git a/jslib b/jslib
index f29afc7c..f80e8946 160000
--- a/jslib
+++ b/jslib
@@ -1 +1 @@
-Subproject commit f29afc7cf7bbd514d07c0354b3c45491dcae3c14
+Subproject commit f80e89465ffc004705d2941301c0ffb6bfd71d1a

**Comment by Arathen** [comment]
Since this has been moved, is there a solution? Or is there more to it?

**Comment by BlackDex** [comment]
Please check the Dockerfile or package_web_vault.sh script on what the steps are, it could be that you mised something.

**Reply by rapkin61** [reply]
I followed exactly the steps described here:

https://github.com/dani-garcia/vaultwarden/wiki/Building-binary#install-the-web-vault

**Reply by BlackDex** [reply]
I think that page needs some updating. So please check those files.

**Reply by rapkin61** [reply]
hmmm ... where can I find that script? Docker is not my passion ...

**Reply by BlackDex** [reply]
It is in the bw_web_vault repo

**Reply by rapkin61** [reply]
ah ... that works with VAULT_VERSION=9d42a565dbef29c5c361dbe5b3db88d765d65a98

I suppose we can ignore all that warnings:


...
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/webpack-dev-server/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/watchpack/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/glob-watcher/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.1.3 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 1319 packages from 638 contributors and audited 1524 packages in 19.804s
found 688 vulnerabilities (659 low, 11 moderate, 18 high)
  run `npm audit fix` to fix them, or `npm audit` for details
**[~/web-vault]$ npm audit fix**
npm WARN deprecated chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.2.7 (node_modules/watchpack-chokidar2/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN @angular/animations@9.1.12 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN @angular/cdk@9.2.4 requires a peer of tslib@^1.9.0 but none is installed. You must install peer dependencies yourself.
npm WARN @angular/common@9.1.12 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN @angular/compiler@9.1.12 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN @angular/compiler-cli@9.1.12 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN @angular/core@9.1.12 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN @angular/forms@9.1.12 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN @angular/platform-browser@9.1.12 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN @angular/platform-browser-dynamic@9.1.12 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN @angular/router@9.1.12 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN angular2-toaster@8.0.0 requires a peer of @angular/common@^8.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN angular2-toaster@8.0.0 requires a peer of @angular/compiler@^8.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN angular2-toaster@8.0.0 requires a peer of @angular/core@^8.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN angulartics2@9.1.0 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN bootstrap@4.3.1 requires a peer of popper.js@^1.14.7 but none is installed. You must install peer dependencies yourself.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.1.3 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/webpack-dev-server/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/glob-watcher/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

+ jquery@3.6.0
+ node-forge@0.10.0
added 17 packages from 11 contributors, removed 6 packages, updated 38 packages and moved 4 packages in 10.836s
fixed 664 of 688 vulnerabilities in 1524 scanned packages
  14 vulnerabilities required manual review and could not be updated
  6 package updates for 10 vulns involved breaking changes
  (use `npm audit fix --force` to install breaking changes; or refer to `npm audit` for steps to fix these manually)

**Comment by jjlin** [comment]
I've updated the build procedure at https://github.com/dani-garcia/vaultwarden/wiki/Building-binary#install-the-web-vault

**Comment by Arathen** [comment]
Thanks for this. The updated build procedure posted by @jjlin above works well. Much appreciated.

**Comment by xiki808** [comment]
I'm getting another error when applying patch v2.20.1 on bitwarden web v2.20.3:

```error: patch failed: package-lock.json:7836
error: package-lock.json: patch does not apply
error: patch failed: src/services/webPlatformUtils.service.ts:212
error: src/services/webPlatformUtils.service.ts: patch does not apply

**Reply by dani-garcia** [reply]
there were some changes in the patched code, I've added a separate patch for v2.20.3 now and it should work again

**Reply by xiki808** [reply]
Thanks will be checking it soon!

**Reply by xiki808** [reply]
Works!

**Comment by Nyxtorm** [comment]
Hello,

With Bitwarden Web 2.20.4, I have a new error when applying the patch 2.20.3:

```bash
error: patch failed: src/services/webPlatformUtils.service.ts:249
error: src/services/webPlatformUtils.service.ts: patch does not apply
```

**Reply by Arathen** [reply]
Im also getting this same error now as well.

**Reply by dani-garcia** [reply]
Fixed in https://github.com/dani-garcia/bw_web_builds/commit/3fba0d5c3a8151418ad636643829ffae0680f52e
